src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py

```diff
@@ -42,7 +43,116 @@
 from functional.iterator import ir as itir
 
 
+_vertical_dimension = "unstructured::dim::vertical"
+_horizontal_dimension = "unstructured::dim::horizontal"
+
+
+def _get_domains(closures: Iterable[itir.StencilClosure]) -> Iterable[itir.FunCall]:
+    domains = [c.domain for c in closures]
+    assert all(isinstance(d, itir.FunCall) for d in domains)
+    return domains
+
+
+def _extract_grid_type(domain: itir.FunCall) -> common.GridType:
+    if domain.fun == itir.SymRef(id="cartesian_domain"):
+        return common.GridType.CARTESIAN
+    else:
+        assert domain.fun == itir.SymRef(id="unstructured_domain")
+        return common.GridType.UNSTRUCTURED
+
+
+def _get_gridtype(closures: list[itir.StencilClosure]) -> common.GridType:
+    domains = _get_domains(closures)
+    grid_types = set(_extract_grid_type(d) for d in domains)
+    if len(grid_types) != 1:
+        raise ValueError(
+            f"Found StencilClosures with more than one GridType: {grid_types}. This is currently not supported."
+        )
+    return grid_types.pop()
+
+
+def _name_from_named_range(named_range_call: itir.FunCall) -> str:
+    assert isinstance(named_range_call, itir.FunCall) and named_range_call.fun == itir.SymRef(
+        id="named_range"
+    )
+    return named_range_call.args[0].value
+
+
+def _collect_dimensions_from_domain(
+    closures: Iterable[itir.StencilClosure],
+) -> dict[str, TagDefinition]:
+    domains = _get_domains(closures)
+    offset_definitions = {}
+    for domain in domains:
+        if domain.fun == itir.SymRef(id="cartesian_domain"):
+            for nr in domain.args:
+                dim_name = _name_from_named_range(nr)
+                offset_definitions[dim_name] = TagDefinition(name=Sym(id=dim_name))
+        elif domain.fun == itir.SymRef(id="unstructured_domain"):
+            if len(domain.args) > 2:
+                raise ValueError("unstructured_domain must not have more than 2 arguments.")
+            if len(domain.args) > 0:
+                horizontal_range = domain.args[0]
+                horizontal_name = _name_from_named_range(horizontal_range)
+                offset_definitions[horizontal_name] = TagDefinition(
+                    name=Sym(id=horizontal_name), alias=_horizontal_dimension
```

Comment:
Shouldn’t we take the information if this is the horizontal dimension from the offset provider/dimension definition? This way it seems to be impossible to declare a vertical-only closure (but not sure if this is ever needed).

Reply:
Ideally yes, but the offset_provider doesn't necessarily contain this information if there is no shift in that dimension. Therefore we decided to follow the c++ approach for now. For pure vertical we can still run in Cartesian mode.